### PR TITLE
Validator client refactor

### LINKED ISSUE #336: Validator client to request block production at specific slot
## Description

Once #220 is complete. We should force the validator client to ask for a specific slot to be produced, in case there is timing issues in the grpc or validator client. This may mean the validator client signs a block at the wrong slot. 

Once implemented, we may want the validator client to also ensure the block it is signing is of the correct slot, for added protection.

### LINKED ISSUE #342: Re-implement seperate threading for validator client duties
## Description

I've removed the threading for the request for duties. This is because at the start of an epoch, the duties will update in parallel, and `process_duties` may not see the updated duties when processing. 

The current work-around is to make the duties request sequential, such that `process_duties` always runs after the duties have been updated. 

This is not good, as the core timing thread is linked with the duties request thread. We want these separate, such that we can timeout and kill the duties thread, leaving the core timing and tokio driver thread unaffected. 

The solution is to add a channel, to the tokio runtime, such that the duties thread tells the tokio runtime if duties have been modified (after an epoch or a re-org) and the tokio runtime can then re-process the new duties.

### LINKED ISSUE #604: Onboarding testnet validators
Here I propose a workflow for on-boarding validators to Goerli testnets (but it should also work for mainnet tetsnets).

### Components

There are two software components:

1. the `account_manager` crate, executed via the monolithic `lighthouse` binary.
    - Generates cryptographic keys, initializes validator slashing databases and produces eth1 deposit data.
1.  a single-page web application, served from some sigp domain but also available locally
    - allows users to submit their validator deposit transactions via a browser (e.g., using metamask).

(1) already exists, however it requires additional features to suit this workflow. (2) does not exist yet.

### Workflow

A user wishes to become a validator:

1. The user downloads, builds and installs lighthouse from this repo.
1. The user runs the following command `$ lighthouse account_manager validator new`
    - A new directory is created for that validator in `~/.lighthouse/validators`
    - Deposit and withdrawal BLS keystores are created (see #596).
    - An empty slashing database is created (see #588).
    - An `eth1_tx_data` file is created (this is described later).
1. The user starts the validator (and any other validators in the datadir) by running `$ lighthouse validator_client`
    - the logs of the validator client inform the user that their validator is not yet included in the beacon state (the validator is "unknown").
1. The user goes to the "deposit website" and uploads their `eth1_tx_data` file and uses metamask to create a deposit transaction. It provides some feedback once the deposit is confirmed.
1. The user watches the logs of the validator client (started earlier) to see the validator transition from unknown, to deposited and then to active.

#### The `eth1_tx_data` file

This file is simply the inputs to the [`deposit`](https://github.com/ethereum/eth2.0-specs/blob/03fb0979485a204890053ae0b5dcbbe06c7f1f5c/deposit_contract/contracts/validator_registration.v.py#L67-L70) function on the deposit contract.

It contains a simple, bytes representation of `(pubkey, withdrawal_credentials, signature, deposit_data_root)`.

#### The deposit website

This just needs to be a static, single-page webpage acting as a shim between metamask and the `eth1_tx_data` file.

It only really needs three features:
- A button to upload the `eth1_tx_data`
- A button to create a deposit transaction using the `eth1_tx_data`.
- Some feedback about whether or not the transaction was mined and successful.

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

- Resolves #604
- Resolves #342
- Resolves #336

## Proposed Changes

- Modifies `account_manager` to suit the deposit flow in #604.
- Refactors the validator client to be more use http and be more async.
- Removes gRPC project-wide.
- Adds a maximum skip slots of 128. Without this, it causes lots of problems with high validator counts and skip slots. HTTP requests start to pile up.
- Adds a snazzy new github actions CI.

### CODE REVIEW

File: validator_client/src/config.rs [hunk 1]
```diff
@@ -37,205 +25,78 @@ impl Default for KeySource {
...
-                self.fetch_testing_keypairs(range.clone())?
-            }
-            KeySource::YamlKeypairs(path) => {
-                warn!(
-                    log,
-                    "Private keys are stored insecurely (plain text). Testing use only."
-                );
-
-                load_keypairs_from_yaml(path.to_path_buf())?
+/// Parses the `testnet` CLI subcommand, modifying the `config` based upon the parametes in
```
Comment (paulhauner) [inline validator_client/src/config.rs]: Typo

File: validator_client/src/lib.rs [hunk 2]
```diff
@@ -1,258 +1,276 @@
...
-            "Waiting for next slot";
-            "seconds_to_wait" => duration_to_next_slot.as_secs()
-        );
+        let duties_exit = self
+            .duties_service
+            .start_update_service(&self.context.eth2_config.spec)
+            .map_err(|e| format!("Unable to start duties service: {}", e))?;
 
-        let (exit_signal, exit_fut) = exit_future::signal();
+        self.exit_signals.write().push(duties_exit);
```
Comment (paulhauner) [inline validator_client/src/lib.rs]: Get rid of this RwLock?

File: validator_client/src/validator_directory.rs [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+pub struct ValidatorDirectoryBuilder {
+    directory: Option<PathBuf>,
+    voting_keypair: Option<Keypair>,
+    withdrawal_keypair: Option<Keypair>,
+    amount: Option<u64>,
+    deposit_data: Option<Vec<u8>>,
+    spec: Option<ChainSpec>,
+}
+
+impl ValidatorDirectoryBuilder {
```
Comment (paulhauner) [inline validator_client/src/validator_directory.rs]: Add more comments to these fns

File: validator_client/src/validator_store.rs [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+pub struct ValidatorStore<T, E: EthSpec> {
+    validators: Arc<RwLock<HashMap<PublicKey, ValidatorDirectory>>>,
+    spec: Arc<ChainSpec>,
+    log: Logger,
+    temp_dir: Option<Arc<TempDir>>,
+    fork_service: ForkService<T, E>,
+    _phantom: PhantomData<E>,
+}
+
+impl<T: SlotClock + 'static, E: EthSpec> ValidatorStore<T, E> {
```
Comment (paulhauner) [inline validator_client/src/validator_store.rs]: Add more comments to these fns

File: validator_client/src/validator_directory.rs (hunk 3 above)
Comment (paulhauner) [reply validator_client/src/validator_directory.rs]: I have done this over in another branch that should get merged into Scott's slashing PR:

https://github.com/sigp/lighthouse/tree/paul-slash-protection

I won't add them again here cause it'll cause conflicts.

**Comment by paulhauner** [comment]
All comments have been addressed!

**Comment by Kimtrai2020** [comment]
> ## Issue Addressed
> * Resolves [Onboarding testnet validators #604](https://github.com/sigp/lighthouse/issues/604)
> * Resolves [Re-implement seperate threading for validator client duties #342](https://github.com/sigp/lighthouse/issues/342)
> * Resolves [Validator client to request block production at specific slot #336](https://github.com/sigp/lighthouse/issues/336)
> 
> ## Proposed Changes
> * Modifies `account_manager` to suit the deposit flow in [Onboarding testnet validators #604](https://github.com/sigp/lighthouse/issues/604).
> * Refactors the validator client to be more use http and be more async.
> * Removes gRPC project-wide.
> * Adds a maximum skip slots of 128. Without this, it causes lots of problems with high validator counts and skip slots. HTTP requests start to pile up.
> * Adds a snazzy new github actions CI.

**Pull request merged**